elastic/apm-server
# modelindexer: Scale active indexers based on load

## Motivation/summary

Modifies the `modelindexer` to scale up and down the number of "active" indexers (goroutines pulling out of the internal modelindexer queue), based on consecutive number of flushes up to 25% of the `GOMAXPROCS`.

When N number of full flushes (Reached or exceeded `FlushBytes`) occur, a new active indexer will be created by the scale action, as long as the scaling cooldown has elapsed.

Equally, when N number of timed flushes (due to `FlushInterval`) occur, an active indexer will be scaled down, since not enough load is going through the server to warrant the current number of active indexers.

Active indexer downscaling can also be triggered due to a change in the `GOMAXPROCS`. This is particularly important for containerized or cgroup environments where CPU quotas may be updated "live". In this case, the downscale cooldown is ignored and active indexers are scaled down until until `active <= math.RoundToEven(GOMAXPROCS / 4)`.

When traffic isn't going through an active indexer, a new timer has been introduced to allow completely idle indexers to be scaled back. The idle check interval can be configured via `IdleInterval`.

Scaling is enabled by default, but can be disabled via the configuration option `output.elasticsearch.scaling.enabled: false`.

Last, modifies the default settings for `max_requests` and `flush_bytes` to be `50` and `1MB` respectively. This allows smaller payloads to be sent to Elasticsearch, more available indexers can be used by instances with more processing power, and the indexers are cycled faster, which results in better usage and performance.

## Checklist

~- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)~

## How to test these changes

Run benchmarks on instances `8g` and bigger in ESS and observe CPU utilization reaching ~100%.

## Related issues

Closes #9181
